test/e2e/e2e_test.go

```diff
@@ -0,0 +1,409 @@
+	e2eContainerName                = "do-k8s-e2e"
+)
+
+var (
+	tokenMissingErr = errors.New("token must be specified in DIGITALOCEAN_ACCESS_TOKEN environment variable")
+
+	// De-facto global variables that require initialization at runtime.
+	supportedKubernetesVersions     = []string{"1.16", "1.15", "1.14"}
+	sourceFileDir                   string
+	testdriverDirectoryAbsolutePath string
+	deployScriptPath                string
+
+	// Variables initialized in TestMain that are leveraged by the tests.
+	ctx context.Context
+	p   params
+)
+
+type params struct {
+	long           bool
+	driverImage    string
+	runnerImage    string
+	focus          string
+	kubeconfig     string
+	nameSuffix     string
+	retainClusters bool
+	kubeVersions   []string
+}
+
+func init() {
+	_, filePath, _, _ := runtime.Caller(0)
+	sourceFileDir = filepath.Dir(filePath)
+	testdriverDirectoryAbsolutePath = filepath.Join(sourceFileDir, testdriverDirectoryRelativePath)
+	deployScriptPath = filepath.Join(sourceFileDir, "..", "kubernetes", "deploy", deployScriptName)
+
+	flag.Usage = func() {
+		fmt.Println(`e2e.test runs containerized, external storage end-to-end tests from upstream Kubernetes against a CSI driver.
+
+It supports dynamically creating (and post-test deleting) a DOKS cluster to run a driver-under-test in. The environment
+variable DIGITALOCAN_ACCESS_TOKEN must be set to a DigitalOcean API key for this purpose.
+The cluster will be tagged with "csi-e2e-test" and "version:<sanitized Kubernetes version>" where
+<sanitized Kubernetes version> is the Kubernetes major-minor version replacing dots by dashes, e.g., "version:1-16"
+(DigitalOcean tags must not contain dots).
+The name of a cluster will be "csi-e2e-<sanitized Kubernetes version>-test-<suffix>" where <suffix> is a random
+alphanumeric suffix if not customized through the corresponding command-line flag.
+
+One or more Kubernetes major-minor versions to run tests for may be given. If omitted, tests will be run for all supported
+Kubernetes versions.
+
+External storage end-to-end tests require a Kubernetes version-specific testdriver YAML file to be defined. An error is
+returned if no corresponding file is found for a given Kubernetes release.
+
+Examples:
+
+# Run tests for all supported versions:
+e2e.test
+
+# Run tests for 1.16 only:
+e2e.test 1.16
+
+# Run tests for 1.16 and 1.14 (but not 1.15):
+e2e.test 1.16 1.14
+
+# Create cluster with a specific suffix:
+e2e.test --name-suffix=$(git rev-parse --short HEAD)
+
+# Retain cluster after completion of the tests:
+e2e.test --retain
+
+# Use cluster referenced by kubeconfig file instead of using a dynamic cluster:
+e2e.test --kubeconfig=$HOME/.kube/config
+
+# Use a custom driver image:
+e2e.test --driver-image=timoreimann/do-csi-plugin:dev
+
+# Use a custom end-to-end test runner image:
+e2e.test --runner-image=timoreimann/k8s-e2e-test-runner:latest
+
+Options:`)
+		flag.PrintDefaults()
+	}
+}
+
+func TestMain(m *testing.M) {
```

Comment:
It would be useful to add a `nodeCount` flag for when we are not using the kubeconfig flag to vary how many nodes in the default node pool the created cluster has. This will enable us to take more advantage of the parallelization that the upstream e2e suite offers if desired.

Reply:
That's definitely doable. To fully take advantage of resources, however, we likely want to make both the pool and ginkgo node counts customizable. We also need to keep in mind that there's some risk involved that our bottleneck moves from the worker nodes to the control plane, so those knobs need to be tuned somewhat carefully.

For context, we already parallelize the (parallelizable) e2e storage tests with a ginkgo node count of 10, and they tend to finish in ~7-9 minutes which seems fairly quick IMO.

What I could do is run the tests with something like 10 or 15 worker nodes and a higher ginkgo node count in order to estimate the maximum speed gain. That may help us indicate the usefulness of exposing more knobs.

Reply:
I compared run latencies of the 3 node setup this PR suggests and one where 20 worker nodes are launched and ginkgo's node count is also set to 20. The results were inconsistent: at times, I was able to shave off 2 minutes; other times, the test duration didn't vary from the 3 node setup.

Overall, making this configurable does not seem to provide a guaranteed impact. I suspect that performance variations in server-sided volume operations have a significant impact that we cannot outperform independently.